### PR TITLE
Screen Manager

### PR DESCRIPTION
**Pull request by Mikotey**
# Description

Begins the game in an initial scene with a Screen Manager. The Screen Manager handles data and transitioning between screens, and having multiple scenes open in the case of overlay screens.
The Screen Manager starts by loading the Splash Screen, then the Splash Screen communicates with the Screen Manager to transition to the Main Menu.
Added a confirmation popup screen, and the Settings and Main Menu were modified to test overlay screens.
While modifying Main Menu, I noticed that the MainMenuController didn't serve much function besides kicking off the MainMenuUiManager, so I consolidated the functionalities of the two and removed the MainMenuController

## Related issue(s)

Sort of related to PR: https://github.com/Saplings-Projects/Birthday-2024/pull/28
The Screen Manager should make it easier to fix the issue of navigating to the Level Select after returning from the Main Level scene. Either by separating Level Select into a separate scene and navigating directly, or by passing transition data

## List of changes

-Initial scene with Screen Manager
-placeholder splash screen
-updating scenes with screen logic
-moving settings to separate scene
-adding confirmation popup scene
-confirmation when trying to exit main menu
-consolidating MainMenu state controllers

### CODE REVIEW

File: Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd
Comment (Casulono) [reply Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd:27]: overall I think Puzzle screen should utilize ScreenManager already in this PR

File: Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd
Comment (Casulono) [inline Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd:1]: ScreenLogic is not initalized in main menu when we go back to it from Play or Win state which causes game to crash once you click Settings or Exit

File: Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd
Comment (Casulono) [inline Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd:29]: instead of calling screenManager.ShowConfirmationPopup() it should call for ShowConfirmationPopup() from this ile from line 24

File: Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd
Comment (Casulono) [inline Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd:24]: this function is unused unless you will modify ExitApplication()

**Review by Casulono — CHANGES_REQUESTED** [review]

File: Birthday-2024-Project/Scripts/UI/MainMenu/States/MainMenuUiStartState.gd
Comment (Mikotey) [reply Birthday-2024-Project/Scripts/UI/MainMenu/States/MainMenuUiStartState.gd]: removed

File: Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd
Comment (Mikotey) [reply Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd:27]: Screen Manager applied to Main Level

File: Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd
Comment (Mikotey) [reply Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd:22]: Screen Manager applied to Main Level

File: Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd
Comment (Mikotey) [reply Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd:1]: Navigating back to Main Menu now uses the updated method. Issue is now fixed

File: Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd
Comment (Mikotey) [reply Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd]: Now using the local function version

File: Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd
Comment (Mikotey) [reply Birthday-2024-Project/Scripts/ScreenManagement/ScreenLogic.gd:24]: The functions on ScreenLogic are mostly just acting as a relay to the ScreenManager. While this function is only currently utilized by the ExitApplication function, it is available for use by anything that might need a two-button popup

File: Birthday-2024-Project/Scripts/ScreenManagement/ScreenManager.gd
Comment (Casulono) [inline Birthday-2024-Project/Scripts/ScreenManagement/ScreenManager.gd:9]: this needs to be looked into closer; using ScreenManager to load main_level messes up with pieces in level
image

**Review by Casulono — CHANGES_REQUESTED** [review]

File: Birthday-2024-Project/Scripts/ScreenManagement/ScreenManager.gd
Comment (Mikotey) [reply Birthday-2024-Project/Scripts/ScreenManagement/ScreenManager.gd:9]: I encountered this issue and yet it completely slipped my mind. It's fixed now. Though, the solution re-introduced the issue with aspect ratios wider than 16:9 when on the level screen. Can probably tackle that later

**Review by Casulono — APPROVED** [review]

**Comment by Mikotey** [comment]
Oh shit! I totally missed your comment! Like not even a convenient backpedal. I recall seeing the PR get approved in the afternoon by Casulono. At the time there was no comment. I wanted to hear your feedback on the PR, so I waited till the next evening. For some reason I got it in my head that you would notify in project chat if you added your feedback. I was super distracted watching Biboo Miami and Ririka English streams and passive working without even paying attention. Absolutely my bad

This is still the most recent push to main, so it's still possible to revert it and make changes.

At least for the reasoning behind merging the statemachines for MainMenu is because it was making it difficult to keep track of two state machines. Ultimately, the main menu is just going to be a menu, thus, only going to be UI. Merging to one made it easier to keep track of things, at least for me. And if we weren't going to use MainMenuController, there was little reason to keep the code around cause the scripts are already kinda all over the place.

This change is only to the Main Menu. Looking at the modified files again, I will agree that merging PuzzleMainScreen was rather unnecessary. Just so I could reduce the number of syntax needed to refer to the PuzzleMainScreen elements. Fundamentally, the state machine and the functionalities are the same. Just the main menu now uses 1 state machine, the ui state machine. All other destructive changes are either replacing the temp code with their proper systems and removing the settings window since now there is a globally accessible settings window that can be overlaid on any screen

The plan I had in mind for the Campaign states was to separate them into their own screens. That way the screen manager can just navigate directly to them, instead of having to navigate to main menu. With the transaction data, it could also navigate to main menu and be informed of which state to transition to as the other potential implementation

**Comment by Casulono** [comment]
I agree with having managers simplified for easier maintenance. So all I all to align what should be the shape of the manager and menus:
- UI and states handled by single manager for main menu
- each new screen should be a new state and screen, so screens done in https://discord.com/channels/877085190655205406/1162414864845783161/1197061158696407080 should be reworked to align to this, should it be done in this PR after reverting it or in new PR?
- Puzzle screens should be handled by different manager and reverting merge of files relevant to UI and state might be needed

But in this case should we make manager for manager?

**Comment by BishopJohnson** [comment]
In my experience it's easier to have functionality split up into more classes/scripts so that each one is only in charge of what is necessary. The framework I was going for was a Model-View-Controller type that I've used in other projects that so far worked better than other stuff I've tried in the past and the goal is to divide up the logic to make organization of responsibility easier. For example the `MainMenuController` wouldn't have been responsible for the actual UI of the main menu, just the underlying logic. So if, like in @Casulono's case the main menu needed to start in a particular state, then the controller would handle that and the UI scripts would respond to it. This also makes it easier to iterate on the UI without fear of breaking the game logic since the UI scripts don't directly affect it without going through the controller and ought to make working in parallel with each other and merging changes easier since the functions are divided up into separate classes.

My main concern as I mentioned is that I think this rework wasn't necessary if the goal of it was to make progress on PR #28 (although I am also concerned that breaking changes are occurring without at least some discussion as a team). And as I mentioned in a [comment](https://github.com/Saplings-Projects/Birthday-2024/pull/28#issuecomment-1890799676) on the PR, I believe that the issue could have been resolved by adding class(es) that model the game (namely what level the player was on). And I think that class could be used to pass data between scenes that the controllers and UI can refer to and probably also set it up to have some serialized variables and make progress on game saves. Also by having such a model that exists globally, it would decouple the controllers of each scene (i.e. main menu and puzzle) so that they won't need to refer to each other and can be agnostic of how each other work.

**Comment by Casulono** [comment]
@Mikotey in this case, what would you say on revert of this commit and applying classes mentioned by Bishop?

**Comment by Mikotey** [comment]
> In my experience it's easier to have functionality split up

Makes sense. That model was useful when making changes to the main level for editor mode.

Some questions on PR https://github.com/Saplings-Projects/Birthday-2024/pull/28 and code clean up:
With the implementation of the Screen Manager, I had planned to implement the animated transition there. So when a screen transitions to another screen, it would perform the page turn animation. Thus, with https://github.com/Saplings-Projects/Birthday-2024/pull/28 , the expectation would be that campaign select, level select, and other menu navigations would be turned to screens instead of states so they can utilize this feature. In this case, the Main Menu will likely remain a single-state.
In this scenario, with Main Menu potentially not needing a state machine at all, it could be considered deprecated code. Would it be preferred to clean it up or leave it? Leaving it has it available for use in the future since it's already implemented, but also has the drawback of reducing code readability since there is under-used code around.

> @Mikotey in this case, what would you say on revert of this commit and applying classes mentioned by Bishop?

Reverting the changes shouldn't be hard. It was more like an after implementation clean up.

> (although I am also concerned that breaking changes are occurring without at least some discussion as a team)

If you see anything as a red flag, do speak up! It's important we have discussions. I didn't mean to push this forward as if to dismiss any discussion. I just was stupid and not paying attention. Is it only in reviews that it blocks the PR until resolved? At least if we do that, we can use that as a failsafe to prevent PRs until any concerns are addressed.

> And as I mentioned in a https://github.com/Saplings-Projects/Birthday-2024/pull/28#issuecomment-1890799676 on the PR, I believe that the issue could have been resolved by adding class(es) that model the game (namely what level the player was on). And I think that class could be used to pass data between scenes that the controllers and UI can refer to and probably also set it up to have some serialized variables and make progress on game saves. Also by having such a model that exists globally, it would decouple the controllers of each scene (i.e. main menu and puzzle) so that they won't need to refer to each other and can be agnostic of how each other work.

If I understand this correctly, so a state machine that represents every screen?

As it stands, the Screen Manager accomplishes most of these tasks. It stores the name of the last screen it was on and passes a dictionary that can be used to pass data between screens when transitioning. This dictionary is stored on the Screen logic script so any node and UI can view it. Though, the main benefit of the Screen Manager is the ability to "stack" screens. This way, sub-screens like the Settings popup or other generic popups like Confirmation popup can be overlaid on any screen with minimal implementation.

**Comment by Casulono** [comment]
Ok, we can take these steps in this case: if the current implementation is helping a lot with planned animations for screen changes, let's leave them. The state machine code can be commented out and moved as a reference to the end of the file with a description.
About avoiding merging changes and discussion: it will slow down merging but after review, if everything is okay, the reviewer should leave a comment that everything looks good before actually approving it leaving 24 hours for any comments before merge. If no one has anything to say, then the merge proceeds. If there are some doubts/questions/etc., merge and approval should take place only after everything is clear.
Does both idea sound good?

**Comment by Mikotey** [comment]
I'll undo the changes since this was my mistake. The Screen Manager is not incompatible with the state machine, so the clean-up can be undone and still function. It'll also be an interesting experience trying to undo things after the fact.

Though, given how Bishop hasn't said anything after so many days have passed, I doubt he is interested in returning to the project even if we do undo the changes. I just hope this hasn't soured his experience working on future projects or with the sapling community.

This is my first time working on an open source project and with people over the internet. Realistically, you're not gonna review every PR for any destructive actions that might concern you. In larger group sizes, that's impossible. Thus, to reduce the risk of undoing another's work or introducing incompatibilities, perhaps code clean up is not a thing for open source projects. Or maybe there's dedicated code clean-up PRs, so people know which ones are most likely to concern them.

> About avoiding merging changes and discussion: it will slow down merging but after review, if everything is okay, the reviewer should leave a comment that everything looks good before actually approving it leaving 24 hours for any comments before merge. If no one has anything to say, then the merge proceeds. If there are some doubts/questions/etc., merge and approval should take place only after everything is clear.

This could create a queue of PRs awaiting approval. Given the size of this project and the due date fast approaching, I think it would be better to not add this wait time. I'll just be more careful of destructive actions in general.

**Pull request merged**
